### PR TITLE
Fix Account Security Page

### PR DESCRIPTION
**Pull request by kopancek**
Account security page and password page have the same functionality, but based on some conditions we were either showing one or the other. This does not make sense, so in this PR we are making the Account Security page the default in favor of password page.

Also, the account security page does not show results for non site admin users in case of SAML or OpenIDConnect auth provider is configured. For that I have added a new field `publicAccountData` on the resolver. By doing this, we will also be moving [some of the complex business logic from UI to backend](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@2d51686d6371d358067a0631c75f02c16905bf80/-/blob/client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx?L87)

## TODO

- [X] Refactor the UI to use the new endpoint
- [X] Cleanup the unneeded business logic from UI
- [X] Refactor the backend to make the code more readable and easier to understand
- [x] Unit tests
- [x] Test locally with SAML and openid connect providers
- [x] Handle bitbucket now that it's an auth provider
- [x] Implement a redirect from deleted /password route to the account security page
- [ ] Fix the buildkite build so that we can actually merge it

## Screenshot
Screenshot 2023-01-17 at 09 37 19

## Test plan

Tested locally by running the following graphql query:
```
query MinExternalAccounts($username: String!) {
        user(username: $username) {
            externalAccounts {
                nodes {
                    id
                    serviceID
                    serviceType
                    publicAccountData {
                      displayName
                      login
                      url
                    }
                }
            }
        }
    }
```

Also added some basic unit tests.

### CODE REVIEW

File: cmd/frontend/graphqlbackend/external_account.go [hunk 1]
```diff
@@ -91,3 +92,186 @@ func (r *externalAccountResolver) AccountData(ctx context.Context) (*JSONValue,
...
+				}
+			}
+			return assigned
+		}
+		if !assignIfDefined([]*string{m.UserClaims.Username, m.UserClaims.GivenName, m.UserClaims.Name, m.UserInfo.Email}) {
+			return nil, nil
+		}
+		return &extsvc.PublicAccountData{
+			UserName: username,
+		}, nil
```
Comment (mrnugget) [inline cmd/frontend/graphqlbackend/external_account.go:257]:
```suggestion
		var username string
		candidates := []*string{m.UserClaims.Username, m.UserClaims.GivenName, m.UserClaims.Name, m.UserInfo.Email}
		for _, value := range candidates {
			if value != nil && *value != "" {
				username = *value
				break
			}
		}
		if username == "" {
			return nil, nil
		}
		return &extsvc.PublicAccountData{
			UserName: &username,
		}, nil
```

File: cmd/frontend/graphqlbackend/external_account.go [hunk 2]
```diff
@@ -91,3 +92,186 @@ func (r *externalAccountResolver) AccountData(ctx context.Context) (*JSONValue,
...
+	Username *string `json:"name,omitempty"`
+	Login    *string `json:"login,omitempty"`
+	URL      *string `json:"html_url,omitempty"`
+}
+
+type GitLabAccountData struct {
+	Username *string `json:"name,omitempty"`
+	Login    *string `json:"username,omitempty"`
+	URL      *string `json:"web_url,omitempty"`
+}
```
Comment (pjlast) [inline cmd/frontend/graphqlbackend/external_account.go:139]: So for these ones I think we can reuse the existing methods.

For GitHub there is `GetExternalAccountData` in `internal/extsvc/github/common.go`
For GitLab there is `GetExternalAccountData` in `internal/extsvc/gitlab/user.go`

This should reduce some of the boiler plate a bit. Then you don't need these structs

File: cmd/frontend/graphqlbackend/external_account.go [hunk 3]
```diff
@@ -91,3 +92,186 @@ func (r *externalAccountResolver) AccountData(ctx context.Context) (*JSONValue,
...
+		m := &GitLabAccountData{}
+		err := json.Unmarshal(val, &m)
+		if err != nil {
+			return nil, err
+		}
+		return &extsvc.PublicAccountData{
+			UserName: m.Username,
+			Login:    m.Login,
+			URL:      m.URL,
+		}, nil
```
Comment (pjlast) [inline cmd/frontend/graphqlbackend/external_account.go:204]: See comment above, I think we can use the `GetExternalAccountData` methods for GitHub and GitLab here

File: cmd/frontend/graphqlbackend/external_account_data_resolver.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+	Username1 *SAMLAttribute `json:"username,omitempty"`
+	Username2 *SAMLAttribute `json:"http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name,omitempty"`
+	Email1    *SAMLAttribute `json:"emailaddress,omitempty"`
+	Email2    *SAMLAttribute `json:"http://schemas.xmlsoap.org/claims/EmailAddress,omitempty"`
+	Email3    *SAMLAttribute `json:"http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress,omitempty"`
+}
+
+type SAMLAttribute struct {
+	Values []*SAMLValue `json:"Values"`
+}
```
Comment (pjlast) [inline cmd/frontend/graphqlbackend/external_account_data_resolver.go:44]: For all of these you can omit the `*` and just have Go default to the empty values (i.e. `[]` and `""`) if they're not present. Since you're checking for the empty values below anyways after confirming they're not nil. This way you can remove all of the nil checks

File: cmd/frontend/graphqlbackend/external_account_data_resolver.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+				username = c.Values[0].Value
+				break
+			}
+		}
+		if username == "" {
+			return nil, nil
+		}
+		return &extsvc.PublicAccountData{
+			Username: &username,
+		}, nil
```
Comment (pjlast) [inline cmd/frontend/graphqlbackend/external_account_data_resolver.go:110]: If you remove the nil possibilities, then this becomes:

```suggestion
		v := data.Values

		candidates := []SAMLAttribute{v.Nickname, v.Login, v.Username1, v.Username2, v.Email1, v.Email2, v.Email3}
		for _, c := range candidates {
			if len(c.Values) > 0 && c.Values[0].Value != "" {
				return &extsvc.PublicAccountData{
					Username: &c.Values[0].Value,
				}, nil
			}
		}
		return nil, nil
```

Comment (pjlast) [reply cmd/frontend/graphqlbackend/external_account_data_resolver.go:110]: There's also a way to do this by unmarshaling it into a `map[string]SAMLAttribute` instead.

But I don't know if one is necessarily better than the other one. It's maybe a bit less clunky? Not sure

Comment (pjlast) [reply cmd/frontend/graphqlbackend/external_account_data_resolver.go:110]: I think given that we're unmarshaling using a list of candidate strings, it might be better to unmarshal using the map

File: cmd/frontend/graphqlbackend/external_account_data_resolver.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+	case "openidconnect":
+		data, err := encryption.DecryptJSON[OpenIDConnectAccountData](ctx, account.Data)
+		if err != nil {
+			return nil, err
+		}
+
+		var username string
+		candidates := []*string{data.UserClaims.Username, data.UserClaims.GivenName, data.UserClaims.Name, data.UserInfo.Email}
```
Comment (pjlast) [inline cmd/frontend/graphqlbackend/external_account_data_resolver.go:118]: Similar here, I think unmarshaling using a map would make more sense

File: cmd/frontend/graphqlbackend/external_account_data_resolver.go (hunk 4 above)
Comment (kopancek) [reply cmd/frontend/graphqlbackend/external_account_data_resolver.go:44]: That's a good idea. I originally defined them as pointers to keep the struct behavior same as we had on the Typescript code. It also seemed a bit cleaner in terms of interface definition. But maybe not :)

File: cmd/frontend/graphqlbackend/external_account_data_resolver.go (hunk 5 above)
Comment (kopancek) [reply cmd/frontend/graphqlbackend/external_account_data_resolver.go:110]: I think I'm going to move this code to the auth provider itself, so that they have a method called `GetExternalAccountData` that will hide all the complexity.

File: client/web/src/user/settings/routes.tsx [hunk 7]
```diff
@@ -49,12 +48,6 @@ export const userSettingsAreaRoutes: readonly UserSettingsAreaRoute[] = [
         exact: true,
         render: lazyComponent(() => import('./profile/UserSettingsProfilePage'), 'UserSettingsProfilePage'),
     },
-    {
-        path: '/password',
```
Comment (mrnugget) [inline client/web/src/user/settings/routes.tsx:53]: Heads up: this test will probably break: https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/integration/profile.test.ts?L166:75&subtree=true

File: client/web/src/user/settings/sidebaritems.ts [hunk 8]
```diff
@@ -29,12 +21,10 @@ export const userSettingsSideBarItems: UserSettingsSidebarItems = [
         to: '/tokens',
         condition: () => window.context.accessTokensAllow !== 'none',
     },
-    //  future GA Cloud nav items
     {
         label: 'Account security',
```
Comment (mrnugget) [inline client/web/src/user/settings/sidebaritems.ts:25]: Should we rename this to `Password & Account Security`? (Just an idea)

File: enterprise/cmd/frontend/internal/auth/oauth/provider.go [hunk 9]
```diff
@@ -73,6 +76,18 @@ func (p *Provider) Refresh(ctx context.Context) error {
...
+func (p *Provider) ExternalAccountInfo(ctx context.Context, account extsvc.Account) (*extsvc.PublicAccountData, error) {
+	switch account.ServiceType {
+	case extsvc.TypeGitHub:
+		return github.GetPublicExternalAccountData(ctx, &account.AccountData)
+	case extsvc.TypeGitLab:
+		return gitlab.GetPublicExternalAccountData(ctx, &account.AccountData)
+	}
+
+	// TODO: add bitbucket cloud when the bitbucket oauth provider is merged
+	return nil, errors.Errorf("Not implemented, sourcegraph only supports github and gitlab for OAuth at the moment")
```
Comment (mrnugget) [inline enterprise/cmd/frontend/internal/auth/oauth/provider.go:88]:
```suggestion
	return nil, errors.Errorf("Sourcegraph currently only supports GitHub and GitLab as OAuth providers")
```

**Review by mrnugget — APPROVED** [review]
Nice, that's pretty clean!

File: client/web/src/user/settings/routes.tsx (hunk 7 above)
Comment (eseliger) [reply client/web/src/user/settings/routes.tsx:53]: should we just redirect instead?

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 10]
```diff
@@ -5890,6 +5890,18 @@ type ExternalAccount implements Node {
...
+    may query this field.
+    """
+    publicAccountData: PublicExternalAccountData
+}
+
+type PublicExternalAccountData {
+    displayName: String
+    login: String
+    url: String
 }
```
Comment (eseliger) [inline cmd/frontend/graphqlbackend/schema.graphql:5905]: doc strings please :)

**Review by eseliger — APPROVED** [review]

File: client/web/src/user/settings/routes.tsx (hunk 7 above)
Comment (kopancek) [reply client/web/src/user/settings/routes.tsx:53]: I tried doing a redirect and failed. Maybe we can create a followup PR and pair on it?

Comment (mrnugget) [reply client/web/src/user/settings/routes.tsx:53]: How'd you fail? What was the problem? It should "just" be a `render: <Redirect />` basically

See: https://sourcegraph.sourcegraph.com/search?q=context%3Asourcegraph+redirect+file%3Aroutes&groupBy=repo

**Comment by mrnugget** [comment]
Heads up: I don't think this should be in here? 
screenshot_2023-01-16_18 35 19@2x

File: client/web/src/user/settings/cloud-ga.ts [hunk 11]
```diff
@@ -17,14 +17,6 @@ export interface Owner {
 
 export const externalServiceUserMode = (): 'disabled' | 'public' | 'all' | 'unknown' => 'disabled'
```
Comment (pjlast) [inline client/web/src/user/settings/cloud-ga.ts:18]: With the removal of the functions below, I think this function is no longer used either

**Review by pjlast — APPROVED** [review]
i-like-it-a-lot-jim-carrey

Just need to check whatsup with that `schema.ts` file that @mrnugget pointed out 🙏

File: client/web/src/user/settings/routes.tsx (hunk 7 above)
Comment (kopancek) [reply client/web/src/user/settings/routes.tsx:53]: Yes, exactly that was not working properly - it was not redirecting, but showing a 404 instead. Let me try again...

File: client/web/src/user/settings/sidebaritems.ts (hunk 8 above)
Comment (kopancek) [reply client/web/src/user/settings/sidebaritems.ts:25]: I would rather keep the name change out of scope of this PR. Longer names might not fit into the sidebar properly and this IMO would benefit from a designer input.

**Comment by kopancek** [comment]
> Heads up: I don't think this should be in here? screenshot_2023-01-16_18 35 19@2x

No idea where that came from 🙄  Removed 🗑️

File: client/web/src/user/settings/routes.tsx (hunk 7 above)
Comment (kopancek) [reply client/web/src/user/settings/routes.tsx:53]: OK, found out why it didn't work previously. I used absolute path before, now with relative it works. The path contains the username of the user and this seems like the easiest implementation of the redirect in the end.

**Comment by unknwon** [comment]
Looks awesome!
